File: app/src/main/java/com/test/homeautomation/screens/main/MainActivity.java
package com.test.homeautomation.screens.main;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;
import com.test.homeautomation.R;
import com.test.homeautomation.api.ApiUtils;
import com.test.homeautomation.models.Device;
import com.test.homeautomation.screens.add_device.AddDeviceActivity;
import com.test.homeautomation.screens.main.recycler.RecyclerViewAdapter;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private ArrayList<com.test.homeautomation.models.Device> devices = new ArrayList<>();
    RecyclerViewAdapter adapter;

    SocketsHelper socketsHelper;

    TextView temperature;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        RecyclerView rv = findViewById(R.id.my_recycler_view);
        adapter = new RecyclerViewAdapter(devices);

        rv.setAdapter(adapter);
        rv.setLayoutManager(new LinearLayoutManager(this));

        temperature = findViewById(R.id.main_temperature);

        socketsHelper = new SocketsHelper(
                new Function1<Double, Unit> () {
                    @SuppressLint("SetTextI18n")
                    @Override
                    public Unit invoke(Double newTemperature) {
                        temperature.setText(newTemperature.toString() + " ℃");
                        return Unit.INSTANCE;
                    }
                }
        );

    }

    @Override
    protected  void onStart(){
        super.onStart();

        Call<List<Device>> getDS = ApiUtils.getApiService().getDevices();
        getDS.enqueue(new Callback<List<Device>>() {
            @Override
            public void onResponse(Call<List<Device>>
                                           call, Response<List<Device>> response) {
                if (!response.isSuccessful()) {
                    Toast.makeText(
                            getApplicationContext(),
                            "Unable to call API",
                            Toast.LENGTH_SHORT
                    ).show();

                }
                adapter.setDevices(response.body());
            }

            @Override
            public void onFailure(Call<List<Device>> call, Throwable t) {
                Toast.makeText(getApplicationContext(),
                        "Unable to call API", Toast.LENGTH_SHORT).show();
                call.cancel();
            }
        });
    }

    public void add_device_popup(View view) {
        startActivity(new Intent(MainActivity.this, AddDeviceActivity.class));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        socketsHelper.close();
    }
}

File: app/src/main/java/com/test/homeautomation/screens/main/MainActivity.kt
package com.test.homeautomation.screens.main

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import com.test.homeautomation.R
import com.test.homeautomation.api.ApiManager
import com.test.homeautomation.models.Device
import com.test.homeautomation.models.TemperatureResponse
import com.test.homeautomation.screens.add_device.AddDeviceActivity
import com.test.homeautomation.screens.main.recycler.RecyclerViewAdapter
import com.test.homeautomation.utils.MessageHelper
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*
import kotlin.properties.Delegates

class MainActivity : AppCompatActivity() {

    private val devices = ArrayList<Device>()
    internal var adapter: RecyclerViewAdapter = RecyclerViewAdapter(devices)

    private lateinit var socketsHelper: SocketsHelper

    var waitingCalls: Int by Delegates.observable(0) { _, _, newValue ->
        main_refresh.isRefreshing = (newValue > 0)
    }

    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        main_add_device.setOnClickListener {
            startActivity(Intent(this@MainActivity, AddDeviceActivity::class.java))
        }

        main_recycler.adapter = adapter
        main_recycler.layoutManager = LinearLayoutManager(this)

        socketsHelper = SocketsHelper { newTemperature ->
            main_temperature.text = "$newTemperature ℃"
            Unit
        }

        main_refresh.setOnRefreshListener { populate() }
    }

    override fun onStart() {
        super.onStart()
        populate()
    }

    override fun onDestroy() {
        super.onDestroy()
        socketsHelper.close()
    }

    // helper function to populate the screen (devices and temperature)
    private fun populate() {
        waitingCalls = 2

        // update the devices and their states
        ApiManager.service.devices.enqueue(object : Callback<List<Device>> {
            override fun onResponse(call: Call<List<Device>>, response: Response<List<Device>>) {
                if (!response.isSuccessful) {
                    MessageHelper.showToast("Error")
                    return
                }

                adapter.setDevices(response.body()!!)
                --waitingCalls
            }

            override fun onFailure(call: Call<List<Device>>, t: Throwable) =
                MessageHelper.showToast("Unable to call API")
        })

        // update the temperature
        ApiManager.service.temperature.enqueue(object : Callback<TemperatureResponse> {
            @SuppressLint("SetTextI18n")
            override fun onResponse(call: Call<TemperatureResponse>, response: Response<TemperatureResponse>) {
                if (!response.isSuccessful) {
                    MessageHelper.showToast("Error")
                    return
                }

                main_temperature.text = String.format("%.2f ℃", response.body()!!.temperature)
                --waitingCalls
            }

            override fun onFailure(call: Call<TemperatureResponse>, t: Throwable) =
                    MessageHelper.showToast("Unable to call API")
        })
    }
}






